File: 2326-spiral-matrix-iv/2326-spiral-matrix-iv.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def spiralMatrix(self, m: int, n: int, head: Optional[ListNode]) -> List[List[int]]:
        grid = [[-1] * n for _ in range(m)]
        
        rs, rl = 0, m - 1
        cs, cl = 0, n - 1
        
        i, j = 0, 0
        
        curr = head
        
        while curr:
            for j in range(cs, cl + 1):
                if curr:
                    grid[rs][j] = curr.val
                    curr = curr.next
            rs += 1

            for i in range(rs, rl + 1):
                if curr:
                    grid[i][cl] = curr.val
                    curr = curr.next
            cl -= 1

            for j in range(cl, cs - 1, -1):
                if curr:
                    grid[rl][j] = curr.val
                    curr = curr.next
            rl -= 1

            for i in range(rl, rs - 1, -1):
                if curr:
                    grid[i][cs] = curr.val
                    curr = curr.next
            cs += 1
            
        return grid

File: 2326-spiral-matrix-iv/2326-spiral-matrix-iv.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int[][] spiralMatrix(int m, int n, ListNode head) {
        int[][] arr = new int[m][n];
        
        for (int i=0; i < m; i++){
            for (int j=0; j < n; j++){
                arr[i][j] = -1;
            }
        }
        
        ListNode curr = head;
        int cs = 0;
        int ce = n - 1;
        int rs = 0;
        int re = m - 1;
        
        while (curr != null){
            for (int i=cs; i <= ce && curr != null; i++){
                arr[rs][i] = curr.val;
                curr = curr.next;
            }rs++;
            
            for (int i=rs; i <= re && curr != null; i++){
                arr[i][ce] = curr.val;
                curr = curr.next;
            }ce--;
            
            for (int i=ce; i >= cs && curr != null; i--){
                arr[re][i] = curr.val;
                curr = curr.next;
            }re--;
            
            for (int i=re; i >= rs && curr != null; i--){
                arr[i][cs] = curr.val;
                curr = curr.next;
            }cs++;
        }
        
        return arr;
    }
}